src/interface.jl

```diff
@@ -21,7 +21,7 @@ function setup(rule, x; seen = Base.IdSet())
   end
 end
 
-subtract!(x, x̄) = iswriteable(x) ? (x .= x .- x̄) : (x .- x̄)
+subtract!(x, x̄) = iswriteable(x) ? (x .= x .- x̄) : eltype(x).(x .- x̄)
```

Comment:
```suggestion
subtract!(x, x̄) = iswriteable(x) ? (x .= x .- x̄) : convert(typeof(x), x .- x̄)
```
Are there any notable cases where this fails? `@less` tells me that it'll take a simpler path for StaticArrays.

Reply:
I get:
```julia
julia> using StaticArrays

julia> x = @SVector rand(Float32, 100); xb = @SVector rand(100);

julia> @btime Float32.($x .- $xb);
  min 21.919 ns, mean 22.044 ns (0 allocations)

julia> @btime convert.(Float32, $x .- $xb);
  min 43.392 ns, mean 44.969 ns (1 allocation, 16 bytes)
```

Reply:
Oh now I see you don't write eltype. I think that's worse, it won't fuse:
```
julia> @btime convert(typeof($x), $x .- $xb);
  min 24.682 ns, mean 24.869 ns (0 allocations)

julia> x = rand(Float32, 1000); xb = rand(1000); xnew = rand(Float32, 1000);

julia> @btime $xnew .= Float32.($x .- $xb);
  min 287.054 ns, mean 289.051 ns (0 allocations)

julia> @btime $xnew .= convert.(Float32, $x .- $xb);
  min 287.018 ns, mean 287.941 ns (0 allocations)

julia> @btime xnew .= convert(typeof($x), $x .- $xb);
  min 785.714 ns, mean 2.957 μs (3 allocations, 12.02 KiB)
```

Reply:
I was thinking specifically of https://github.com/JuliaGPU/CUDA.jl/issues/261 and how `T.(::AbstractArray{T})` allocates for non-static arrays, but the first is fixed and the second is irrelevant because of fusion.